pkg/sql/sql.go

```diff
@@ -352,6 +352,14 @@ func (c *Client) SetSystemVariables(ctx context.Context, keyVal map[string]strin
 	return nil
 }
 
+func (c *Client) LockTablesWithReadLock(ctx context.Context) error {
+	return c.Exec(ctx, "FLUSH TABLES WITH READ LOCK;")
```

Comment:
my concern here is that once this connection is terminated it unlocks the tables automatically. So we should make sure this connection is active until the failover is complete.

running the following command locks and unlocks tables automatically,

```bash
mariadb -uroot -p${MARIADB_ROOT_PASSWORD} -e 'FLUSH TABLES WITH READ LOCK'
```

Reply:
That's right, I completely forgot ththat the tables are unlocked after the connection is closed. In our case, the connections last one reconciliation cycle, so it should be fine. See:

https://github.com/mariadb-operator/mariadb-operator/blob/0fa82f1aa5a55e6a0eea2625482519cef7ab1ec3/pkg/controller/replication/controller.go#L99

https://github.com/mariadb-operator/mariadb-operator/blob/0fa82f1aa5a55e6a0eea2625482519cef7ab1ec3/pkg/controller/replication/controller.go#L123

I will delete the unlock method, as tables are implicitly unlocked 👍🏻

Reply:
Hmmm, maybe the unlock is required:
- https://github.com/mariadb-operator/mariadb-operator/actions/runs/7273373846/job/19817219567?pr=311

Will take a closer look later

Reply:
Second attempt, this time explicitly unlocking the primary:
- https://github.com/mariadb-operator/mariadb-operator/actions/runs/7297147476/job/19885960383?pr=311

Reply:
Judging by the logs, it seems like we need to unlock the old primary before coverting it into a replica.

Reply:
right, it seems like that. I think this is safe to unlock tables at that phase as it is the last step and the primary service should be using the new primary already

```bash
Error 1192: Can't execute the given command because you have active locked tables or an active transaction
```

Reply:
Third time is the charm:
https://github.com/mariadb-operator/mariadb-operator/actions/runs/7297603999/job/19887135274?pr=311